src/internal/transaction.ts

```diff
@@ -167,9 +169,48 @@ export async function waitForTransaction(args: {
       lastTxn,
     );
   }
+
+  // Make sure indexer is synced with the latest ledger version
+  if (indexerVersionCheck) {
+    try {
+      await waitForLastSuccessIndexerVersionSync({ aptosConfig, ledgerVersion: Number(lastTxn.version) });
+    } catch (_e) {
+      throw new WaitForTransactionError(
+        `Transaction ${txnHash} timed out waiting for indexer to sync with ledger version ${lastTxn.version}`,
+        lastTxn,
+      );
+    }
+  }
+
   return lastTxn;
 }
 
+/**
+ * Waits for the indexer to sync up to the ledgerVersion. Timeout is 3 seconds.
+ */
+async function waitForLastSuccessIndexerVersionSync(args: {
+  aptosConfig: AptosConfig;
+  ledgerVersion: number;
+}): Promise<void> {
+  const { aptosConfig, ledgerVersion } = args;
+  const timeoutMiliseconds = 3000; // 3 seconds
+  const startTime = new Date().getTime();
+  let indexerVersion = -1;
+
+  while (indexerVersion < ledgerVersion) {
+    // check for timeout
+    if (new Date().getTime() - startTime > timeoutMiliseconds) {
+      throw new Error("waitForLastSuccessIndexerVersionSync timeout");
+    }
+
+    // eslint-disable-next-line no-await-in-loop
+    indexerVersion = await getIndexerLastSuccessVersion({ aptosConfig });
+
+    // eslint-disable-next-line no-await-in-loop
+    await sleep(500);
```

Comment:
We will likely want to do an exponential backoff in the future (we should later work on a helper function for it).  Otherwise, you can do no better than 500ms or 1 second here if it doesn't occur immediately.

let's change to 200ms for now.

Reply:
Changed to 200ms